[FLINK-4717] Add CancelJobWithSavepoint

uce opened the pull request:
This builds on top of #2608.

- Adds CancelJobWithSavepoint message, which triggers a savepoint before cancelling the respective job. The periodic checkpoint scheduler is stopped before doing this so that no checkpoints happen between the savepoint and cancellation. Note that users could manually trigger another savepoint while cancel-with-savepoint happens, but I think this is tolerable.

- Adds `-s [targetDirectory]` option to CLI cancel command:
    * Regular cancelling:`bin/flink cancel <jobID>`
    * Cancel with savepoint to default dir: `bin/flink cancel -s <jobID>`
    * Cancel with savepoint to target dir: `bin/flink cancel -s [targetDir] <jobID>`

uce commented:
To help review, the main change is in `JobManager` and `CliFrontend`. Some questions there may be:
- Should we keep the behaviour that the job is not cancelled if the savepoint fails? My reasoning for this was that the user might otherwise accidentally loose state when cancelling the job.
- Does the way that the cancellation is specified on the CLI make sense? Should we keep the `-s` option or add another proper command like `bin/flink cancel-with-savepoint`?

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -581,6 +581,62 @@ class JobManager(
+          case Some((executionGraph, _)) =>
+            // We don't want any checkpoint between the savepoint and cancellation
+            val coord = executionGraph.getCheckpointCoordinator
+            coord.stopCheckpointScheduler()
```
I think it's not enough to simply call `stopCheckpointScheduler`. If I'm not mistaken, then the following could happen: You call `stopCheckpointScheduler` which will try to `cancel` the last `currentPeriodicTrigger`. Now assume that the last `TimerTask` to trigger the next checkpoint has just been triggered but not executed (just before cancelling it). Now the `stopCheckpointScheduler` finishes without the `TimerTask` having completed. Now the `TimerTask` can still trigger a checkpoint even though we've stopped the checkpoint scheduler.

The way to fix this (admittedly academic corner case), is to filter out outdated `TimerTask` calls in the `CheckpointCoordinator` by having a kind of fencing tokens for the trigger checkpoint calls.

uce commented:
Thanks for this review, too. Good catch with the broken `stopCheckpointScheduler`. I will add your proposed fix as a separate commit.

uce commented:
Addressed the issue and found another bug in #2608 that I've fixed in f769e8e. If Travis gives the green light, I will rebase on #2608 and merge this later today.

uce commented:
Local Travis build passed (https://travis-ci.org/uce/flink/builds/167108720). I'm going to rebase and merge this later.
